### ISSUE TITLE
`Partition` Operation - Awkward to use?

### ISSUE THREAD

**Issue by AlexandruGG** [loophp/collection#123]
## Background

I am looking to replace usage of Doctrine Collections with this library in a project, and one of the issues I've come across is the usage of `partition`. Its return type is a bit confusing at first glance and replicating the functionality of the Doctrine Collections method does not seems straightforward for a new user.

## Doctrine Collection

The return type of the [Doctrine Collection Partition](https://github.com/doctrine/collections/blob/1.6.x/lib/Doctrine/Common/Collections/Collection.php#L263) method is very straightforward. It returns an array shape with _exactly two elements_ consisting of the same collection objects with the same  type for the keys and values. Because of this, usage with the PHP [list](https://www.php.net/manual/en/function.list.php) language construct is possible and very convenient:

```PHP
$data = [
    ['id' => 'ABC', 'name' => 'Bob'],
    ['id' => 'BAC', 'name' => 'John'],
    ['id' => 'ABC', 'name' => 'Mike'],
];

[$matches, $nonMatches] = (new ArrayCollection($data))
    ->partition(static fn (int $key, array $entity): bool => $entity['id'] === 'ABC');
```

## Loophp Collection

The return type of the [partition operation](https://github.com/loophp/collection/blob/master/src/Contract/Operation/Partitionable.php#L23) is a collection of lists of arrays of an array shape where the key and value of the elements is included (the nesting is real!).

The first issue is that this operation will always return exactly two elements, but this is not immediately visible from the return type and I'm not sure how this can be fixed because we cannot specify collection shape in the same way as array shape.

The other main issue is the combination of operations needed to achieve the same behaviour as Doctrine Collections: _I just want to be able to get the two separate collections with the values inside them, nothing more._ This is what I've come up with:

```PHP
$partitioned = Collection::fromIterable($data)
    ->partition(static fn(array $entity): bool => $entity['id'] === 'ABC');

$matches = $partitioned->first()->pluck('*.1')->unwrap();
$nonMatches = $partitioned->last()->pluck('*.1')->unwrap();
```

## Possible improvements

A) don't return the keys and the values separately. Rather, they should be together just like in the original collection, before the partition
B) can we reduce the nesting so the `unwrap()` is not necessary?
C) could we return an array shape so that the `list` construct can be used? I realise this is a significant difference, however otherwise we rely on the user "knowing" that this operation will only return two elements in the collection - something that's not currently captured in the type annotations

**Comment by AlexandruGG** [comment]
@drupol hey, let me know please if I'm missing something obvious with this one and if there's an easier way to do what I'm describing 😄

**Comment by drupol** [comment]
Hi Alex,

I never like the implementation of that operation, it is something to improve I think.

There are many ways to fix this, I do not know yet which one is the best.

## Proposal 1

By updating the inner loop as such:

<details>

```php
                    foreach ($iterator as $key => $current) {
                        $result = array_reduce(
                            $callbacks,
                            $reducerCallback($key)($current)($iterator),
                            false
                        );

                        yield (bool) $result => [$key, $current];
                    }
```

</details>

Then, to retrieve the results:

<details>

```php
$collection = Collection::fromIterable(range(0, 15))
    ->partition(
        $isGreaterThanFive,
    );

var_export($collection->all());
```

</details>

The problem is that this won't help. A `boolean` cannot be used as a key in an array,
Thus in order to get the results, you should either use a `foreach` loop and let the user rebuild an array, or add `wrap` operation after the `partition` operation.

## Proposal 2

Update the inner loop as such:

<details>

```php
                    $true = $false = [];

                    foreach ($iterator as $key => $current) {
                        $result = array_reduce(
                            $callbacks,
                            $reducerCallback($key)($current)($iterator),
                            false
                        );

                        $result ?
                            $true[] = [$key, $current] :
                            $false[] = [$key, $current];
                    }

                    yield [$true, $false];
```

</details>

And then, when iterating, you only get one set of results containing at `index 0`, the happy scenario, at `index 1`, the unhappy scenario.

## Proposal 3

Update the inner loop as such:

<details>

```php

                    foreach ($iterator as $key => $current) {
                        yield [
                            (bool) array_reduce(
                                $callbacks,
                                $reducerCallback($key)($current)($iterator),
                                false
                            ),
                            $key,
                            $current
                        ];
                    }
```

</details>

And then, when iterating, you get an array containing 3 elements:

1. The result of the callbacks as bool
2. The key
3. The value

To be honest, I really like proposal 3 and 1 because it's lazy and it doesn't accumulate the results in an array. When I do such things I always think about a huge amount of data in input. If we use a lazy collection library, it's not to fill in arrays with those data, but to process them one-by-one (lazily), this is why I rather avoid proposal 2 (_and the current implementation as it is today_)

WDYT ?

**Comment by AlexandruGG** [comment]
Hey @drupol, thank you for taking the time to look through this and come up with the proposals.

Unfortunately, none of them solve the issues mentioned 😞 . While proposal 3 maintains lazyness I don't think it's what the user would expect. It's more like a `map`, we are just transforming each result into an array of things.

This operation should result in something more straightforward that doesn't require additional processing, because if you think of it at its core, `partition` is just a `filter` where the _passing_ results are put in one group, and the _rejected_ results are put in another group. I think this can still be done in a lazy way.

I've done a bunch of testing and put together a gist here: https://gist.github.com/AlexandruGG/783798339c766454196f7bd2b07a23e9. This is incomplete because I couldn't figure out how to get the exact thing I wanted, I left a comment towards the end.

Effectively the outcome I'm looking for is this:

```PHP
$isEven = static fn(int $item): bool => $item % 2 === 0;
$partitioned = Collection::fromIterable([1, 2, 3])
    ->partition($isEven); 
    // Collection::fromIterable([ 0 => Collection::fromIterable([2]), 1 => Collection::fromIterable([1, 3]) ])
   // note we are still lazy so far

[$even, $odd] = $partitioned->all();
   // now we are not lazy anymore, we've transformed to array
```

Do you think this is possible? How can we modify the gist to achieve this?

**Comment by drupol** [comment]
Hi Alex,

I see.

I don't think this is possible. Doing what you want means that we have to first iterate over ALL the collection before throwing the results, and that is not lazy.
Being lazy means throwing results one item at each iteration, and I don't think this is compatible with what you're trying to do, sadly.

**Comment by AlexandruGG** [comment]
> Hi Alex,
> 
> I see.
> 
> I don't think this is possible. Doing what you want means that we have to first iterate over ALL the collection before throwing the results, and that is not lazy.
> Being lazy means throwing results one item at each iteration, and I don't think this is compatible with what you're trying to do, sadly.

Hmm maybe I'm missing something, but are the current `Filter` and `Diff` operations not already lazy? If so, can we not combine them somehow to achieve this result?

In plain English, the way I think of this process is:
1. process current item, check against the predicate
2. if it passes the predicate, assign to first group of items; if it fails, assign to second group of items

**Comment by AlexandruGG** [comment]
Effectively I could get the results I want from `Partition` by doing two filtering operations. Assuming `Filter` is already lazy, they should be equivalent, right? Following from my previous example:

```PHP
$isEven = static fn(int $item): bool => $item % 2 === 0;
$collection = Collection::fromIterable([1, 2, 3]);
[$even, $odd] = [$collection->filter($isEven), $collection->filter(!$isEven)];
// so far we've just applied the filtering operations and we are left with two separate collections (still lazy)

[$even, $odd] = [$collection->filter($isEven)->all(), $collection->filter(!$isEven)->all()];
// if we use `all()` we can get the final results

**Comment by drupol** [comment]
This is perfectly fine for sure but...

In this case, you are creating 2 collections from one.
One for the happy scenario, one for the unhappy scenario.

What we are trying to achieve here is to make that working using 1 single generator, without using the collection library inside the operation.

Try to create an operation that does that, you'll see.

**Comment by AlexandruGG** [comment]
> This is perfectly fine for sure but...
> 
> In this case, you are creating 2 collections from one.
> One for the happy scenario, one for the unhappy scenario.
> 
> What we are trying to achieve here is to make that working using 1 single generator, without using the collection library inside the operation.
> 
> Try to create an operation that does that, you'll see. The last argument of each Operation is an Iterator, not a Collection... so things needs to be handled in a different way.

Thanks for the clarification, I understand now what you mean. I think ideally `Collection::partition` would create two separate collections like it happens with the [Doctrine ArrayCollection::partition](https://github.com/doctrine/collections/blob/1.6.x/lib/Doctrine/Common/Collections/ArrayCollection.php#L390) (maybe we could do it outside the operation, from the Collection class?).

However, if that's not possible then I propose for the meantime we just tweak the operation a little to make it easier to work with. Here's my proposal:

## Proposal 4

Update the inner loop similar to proposal 2, but associate the keys and values and reduce one level of nesting.

<details>

```php
                    $true = $false = [];

                    foreach ($iterator as $key => $current) {
                        $result = array_reduce(
                            $callbacks,
                            $reducerCallback($key)($current)($iterator),
                            false
                        );

                        $result ?
                            $true[] = [$key => $current] :
                            $false[] = [$key => $current];
                    }

                    yield from [$true, $false];
```
</details>

While this does not solve the lazyness issue, it makes the below possible, which would not have been possible before. Also we can update the type hints to make it clear that only two elements will be returned at the base level.

```php
[$matching, $nonMatching] = $partitioned->all();
```

**Comment by drupol** [comment]
But this is already doable with the current implementation... 

```php


$isGreaterThanFive = static function (int $value): bool {
    return 5 < $value;
};

$input = range(0, 15);

[$true, $false] = Collection::fromIterable($input)
    ->partition($isGreaterThanFive)
    ->all();

```

See my first code snipped in https://github.com/loophp/collection/issues/123#issuecomment-874488389

**Comment by drupol** [comment]
Basically, what you're looking for is this: https://gist.github.com/drupol/ff94497365e9864b610c6c01376f325b right ?

**Comment by AlexandruGG** [comment]
> What if `$key` is not `int|string` then ?
> 
> Anyway, this is almost already doable with the current implementation...
> 
> ```
> $isGreaterThanFive = static function (int $value): bool {
>     return 5 < $value;
> };
> 
> $input = range(0, 15);
> 
> [$true, $false] = Collection::fromIterable($input)
>     ->partition($isGreaterThanFive)
>     ->all();
> ```
> 
> See my first code snipped in [#123 (comment)](https://github.com/loophp/collection/issues/123#issuecomment-874488389)

Hmm I didn't consider the non-array-key issue. Is it possible to use generators there to accommodate that?

I saw your example but it does not return what you might think. That's why I needed to use `pluck` + `unwrap` + `first` / `last` to get what I wanted, as I described. See a simplified example below:

```PHP
$isLetterB = static fn(string $item): bool => $item === 'B';

[$true, $false] = Collection::fromIterable(range('A', 'C'))
    ->partition($isLetterB)
    ->all();

var_dump('matches', $true);
var_dump('no matches', $false);

/*
string(7) "matches"
array(1) {
  [0]=>
  array(2) {
    [0]=>
    int(1)
    [1]=>
    string(1) "B"
  }
}
string(10) "no matches"
array(2) {
  [0]=>
  array(2) {
    [0]=>
    int(0)
    [1]=>
    string(1) "A"
  }
  [1]=>
  array(2) {
    [0]=>
    int(2)
    [1]=>
    string(1) "C"
  }
}
*/
```

**Comment by AlexandruGG** [comment]
> Basically, what you're looking for is this: https://gist.github.com/drupol/ff94497365e9864b610c6c01376f325b right ?
> 
> It can be used as such:
> 
> ```
> $isGreaterThanFive = static function (int $value): bool {
>     return 5 < $value;
> };
> 
> $input = range(0, 15);
> 
> [$true, $false] = Collection::fromIterable($input)
>     ->partition($isGreaterThanFive)
>     ->all();
> ```
> 
> Where `$true` and `$false` are now instances of `CollectionInterface`.

yes, something like that. I tried this, two comments:

- why is the filter change in Collection.php needed? That's not being called during partition, right?
- the results are not collections from what I see, but rather instances of `CallbackFilterIterator`. Only if I do some further processing like `var_dump(iterator_to_array($matching));` I can get the results

I will leave a comment on the gist as well

**Comment by drupol** [comment]
I updated the gist, there was an issue, the mapping was applied on `::filter()`, not on `::partition()`.

Here's how to use it:

```php
$isLetterB = static fn(string $item): bool => $item === 'B';

$partitioned = Collection::fromIterable(array_combine(range('A', 'C'), range('A', 'C')))
    ->partition($isLetterB);

[$happy, $unhappy] = $partitioned->all();

var_dump('matches', $happy->all());
var_dump('no matches', $unhappy->all());
```

**Comment by AlexandruGG** [comment]
> I updated the gist, there was an issue, the mapping was applied on `::filter()`, not on `::partition()`.
> 
> Here's how to use it:
> 
> ```
> $isLetterB = static fn(string $item): bool => $item === 'B';
> 
> $partitioned = Collection::fromIterable(array_combine(range('A', 'C'), range('A', 'C')))
>     ->partition($isLetterB);
> 
> [$happy, $unhappy] = $partitioned->all();
> 
> var_dump('matches', $happy->all());
> var_dump('no matches', $unhappy->all());
> ```

Bingo, this would be really nice. I think it would go in tune with the theme of the library because it returns separate collections.

A couple of questions:
1. what's the effect of `return yield`? I've seen it in a few places and tried testing various things but I still don't see what is the difference between that and just `yield`?

2. do you think the operation would be simplified if we had a [Reject](https://laravel.com/docs/8.x/collections#method-reject) Operation? 😁 

```PHP
// '$not' no longer needed
$happy = Filter::of()(...$callbacks)($iterator);
$unhappy = Reject::of()(...$callbacks)($iterator);

return yield from [$happy, $unhappy];
```

**Comment by drupol** [comment]
> Bingo, this would be really nice. I think it would go in tune with the theme of the library because it returns separate collections.
> 
> A couple of questions:
> 
>     1. what's the effect of `return yield`? I've seen it in a few places and tried testing various things but I still don't see what is the difference between that and just `yield`?

Not really much difference in this particular case. (opinionated) I think it's always cleaner to have a return at the end of the body of a method.

However, an iterator can yield values but it can also return values!

This is something that we should maybe investigate in loophp/collection: https://3v4l.org/0W102
 
>     2. do you think the operation would be simplified if we had a [Reject](https://laravel.com/docs/8.x/collections#method-reject) Operation? grin

Yes, having a new operation `Reject` would definitely make sense.

If we decide to go that way, we will most probably have to rewrite some other operations like `Span`.

**Comment by AlexandruGG** [comment]
Regarding [this change](https://gist.github.com/drupol/ff94497365e9864b610c6c01376f325b#file-partition-diff-L9-L13) from the gist, do you see this as ok to do or does it need to be done in a different way?

**Comment by drupol** [comment]
I don't know yet. Sure it's technically good, but I'm a bit worried about it. Maybe we should open a PR and discuss about it there ?
